### ISSUE TITLE
VS2019 workloads no longer work

### ISSUE THREAD

**Issue by BlythMeister** [jberezanski/ChocolateyPackages#99]
I've just updated to the latest (16.9.x) and now my workloads arguments say they are not valid for the setup.

I get exit code 87.

I'm using `--allWorkloads --includeRecommended --includeOptional`

I have included bootstrapper path to my offline installer, but it seems to call setup.exe on the already installed visual studio installer, rather than the bootstrapper I passed.

**Comment by jberezanski** [comment]
Please provide:
- the exact and complete command line you use
- the Chocolatey log file (`$Env:ChocolateyInstall\Logs\chocolatey.log`)
- the log files generated by VS (`$Env:TEMP\chocolatey\dd_*.log`)

**Comment by BlythMeister** [comment]
Command: 
`choco upgrade visualstudio2019buildtools --source="https://chocolatey.nuget.15below.local/api/odata" --confirm --use-system-powershell --fail-on-standard-error --allow-empty-checksums --allow-downgrade --version="16.9.1.0" --package-parameters="--bootstrapperPath \\BTN-NAS-01\BuildAgents\Chocolatey\VisualStudio\2019\16.9.1.0\vs_BuildTools.exe --noWeb --add Microsoft.VisualStudio.Workload.ManagedDesktopBuildTools --add Microsoft.VisualStudio.Workload.MSBuildTools --add Microsoft.VisualStudio.Workload.NetCoreBuildTools --add Microsoft.VisualStudio.Workload.WebBuildTools --add Microsoft.VisualStudio.Workload.DataBuildTools --add Microsoft.VisualStudio.Workload.AzureBuildTools --includeOptional --includeRecommended --norestart --wait --quiet"`

Files:
[chocolatey.log](https://github.com/jberezanski/ChocolateyPackages/files/6167093/chocolatey.log)
[dd_bootstrapper_20210318171332.log](https://github.com/jberezanski/ChocolateyPackages/files/6167095/dd_bootstrapper_20210318171332.log)
[dd_installer_20210318171333.log](https://github.com/jberezanski/ChocolateyPackages/files/6167096/dd_installer_20210318171333.log)

They key error in dd_installer is:

```
[06b8:0001][2021-03-18T17:13:33] Raw Command line: "C:\Program Files (x86)\Microsoft Visual Studio\Installer\setup.exe"  --layoutPath "\\BTN-NAS-01\BuildAgents\Chocolatey\VisualStudio\2019\16.9.1.0" --in "\\BTN-NAS-01\BuildAgents\Chocolatey\VisualStudio\2019\16.9.1.0\Response.json" update --quiet --installPath "C:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools" --add Microsoft.VisualStudio.Workload.ManagedDesktopBuildTools --add Microsoft.VisualStudio.Workload.MSBuildTools --add Microsoft.VisualStudio.Workload.NetCoreBuildTools --add Microsoft.VisualStudio.Workload.WebBuildTools --add Microsoft.VisualStudio.Workload.DataBuildTools --add Microsoft.VisualStudio.Workload.AzureBuildTools --noWeb --includeRecommended --norestart --includeOptional --locale en-US --activityId "b1034593-1e14-40f4-8375-958c7a5af3a2"
[06b8:0001][2021-03-18T17:13:33] Warning: Closing the installer with exit code 87
[06b8:0007][2021-03-18T17:13:34] Warning: Command line errors:
Option 'includeRecommended' is unknown.
Option 'includeOptional' is unknown.
[06b8:0001][2021-03-18T17:13:34] Warning: Failed to parse the command line arguments: Your parameters contain multiple errors. First error: Option 'includeRecommended' is unknown. Usage: setup.exe update [options]
Command: --add. Usage: The advertised component or workload ID to add
Command: --productKey. Usage: The license product key for the product. Composed of 25 alphanumeric characters either in the format 'xxxxx-xxxxx-xxxxx-xxxxx-xxxxx' or 'xxxxxxxxxxxxxxxxxxxxxxxxx'
Command: --previewUpdate. Usage: Shows information about the update before starting
Command: --focusedUi. Usage: Opens the installer in a focused installation mode
Command: --flight. Usage: The flights to enable, along with their duration e.g. flight1;7d
Command: --updateVersion. Usage: The version of the update to apply
Command: --installPath. Usage: The path to the installation to update
Command: --installLayoutPath. Usage: The path to the layout to use for installation packages for the update
Command: --installCatalogUri. Usage: The URI or path to the catalog to use for the update
Command: --vsix. Usage: The URI or path to additional vsixs to install
Command: --installWhileDownloading. Usage: Enables parallel install and download
Command: --downloadThenInstall. Usage: Downloads all packages before installing
Command: --path. Usage: Sets the custom path using a key=value pair e.g. --path cache=C:\\cachePath
Command: --passive, --p. Usage: The command executes immediately without user interaction, showing UI
Command: --quiet, --q. Usage: The command executes immediately without user interaction, without creating UI
Command: --norestart. Usage: Prevents a quiet/passive operation from rebooting automatically, if required
Command: --channelUri. Usage: The URI or path to the channel manifest to use for updates
Command: --installChannelUri. Usage: The URI or path to the channel manifest to use for the installation
Command: --channelId. Usage: The ID of the channel containing the product to install
Command: --productId. Usage: The ID of the product to install
Command: --layoutPath. Usage: The path to the layout to use for installation packages
Command: --campaign. Usage: The ID for campaign tracking
Command: --activityId. Usage: The ID of the related process that started the installer
Command: --devenvLaunchArguments. Usage: The arguments to pass on to devenv
Command: --in. Usage: The path to the response file to use for the operation
Command: --noUpdateInstaller. Usage: Prevents the installer from automatically updating if an update is required
Command: --cache. Usage: Specifies that payloads should be cached on the system
Command: --nocache. Usage: Specifies that payloads should be deleted from the system
Command: --noWeb. Usage: Prevents the installer from downloading any payloads
Command: --force. Usage: Forces any running, blocking exes to shutdown before starting the operation
Command: --installerFlight. Usage: Flights to enable for the installer, along with their duration e.g. flight1;7d
Command: --theme. Usage: Changes the color theme of the installer
Command: --locale. Usage: Sets the UI locale
[06b8:0001][2021-03-18T17:13:34] Warning: Shutting down the application with exit code 87
```
For some reason its calling `C:\Program Files (x86)\Microsoft Visual Studio\Installer\setup.exe` rather than the bootstrapper path i provided `\\BTN-NAS-01\BuildAgents\Chocolatey\VisualStudio\2019\16.9.1.0\vs_BuildTools.exe`

I have verified that the bootstrapper path does exist and can run that exe manually.

**Comment by BlythMeister** [comment]
I have just run the command again, but with `--debug` to give some extra logging in case this helps :)
[Uploading chocolatey.debug.log…]()

**Comment by BlythMeister** [comment]
I tested by manually calling without the arguements it doesn't like, and it now runs, but says no updates (even though there is)

It seems this new update mechanism from Microsoft is completely flawed!

**Comment by BlythMeister** [comment]
I swapped to using workload package and so far so good.  Maybe the docs on the main package needs updating to say that this is how is should work and calling the base (msbuild) package with some of the arguments no longer work.

**Comment by jberezanski** [comment]
Thank you for the detailed logs and analysis.

Yes, in the VS 16.9 timeframe Microsoft has rather extensively reworked the internals of the VS Installer and little quirks are popping up all over the place.

The docs (https://docs.microsoft.com/en-us/visualstudio/install/use-command-line-parameters-to-install-visual-studio?view=vs-2019) list the possible installer parameters. However, in most cases, there is no information which scenario (install/update/modify/uninstall) a given parameter is valid for.

In particular, the parameters `--allWorkloads`, `--includeRecommended`, `--includeOptional` seem to be valid only in the "install" scenario - when installing a new instance of Visual Studio, not when updating an already installed instance. Now, I don't really remember the behavior of previous versions of the VS Installer if those parameters were provided in an "update" scenario - I mean, they never had an effect during update, but I cannot recall whether their presence caused an error or if they were simply ignored. Clearly, however, the current VS Installer version rejects these parameters with a hard error.

So, the current advice for using the choco packages is:
- when first installing an instance of VS, you can use `--allWorkloads`, `--includeRecommended` and/or `--includeOptional`
  - (but you don't have to - you can add workloads later using the workload packages);
- when updating an already installed instance of VS, those parameters are prohibited (but they are unnecessary anyway, because an update operation will update all already installed workloads);
- for adding new workloads to an already installed instance of VS, use the workload packages.

Now, what remains to be done, perhaps, is to guide users towards proper usage and away from the trap you had fallen into. You mentioned including this info ("these parameters can only be used during first installation") in the description of the base packages ("product packages", as I call them). This would be useful, certainly.

I was also thinking about extending the install script to recognize these parameters and ignore them during update scenario (possibly with a warning to the user). This would support users who enable the choco feature [useRememberedArgumentsForUpgrades](https://docs.chocolatey.org/en-us/configuration#general-1), which - as the docs warn - will be automatically turned on by default in some unspecified future.

**Comment by BlythMeister** [comment]
Thanks for your reply.

I can confirm the old Vs installer would not error on these parameters (that is how my setup was)

However, this has taken me to actually use the workload packages and setup this way which feels like a much nicer way to work anyway.

Personally, I think an update to the docs on your packages to advise on this way to install and not suggesting to use the workloads etc on the main package is a better way to go.

**Comment by jberezanski** [comment]
How about this? https://github.com/jberezanski/ChocolateyPackages/commit/4547114b850f7581bdf2dc5470c636b8c1c9ffc3

**Comment by BlythMeister** [comment]
Perfect!

Had I of seen that, I would have gone "ohhh" and then done it that way from the start 😀

**Comment by jberezanski** [comment]
OK, so I'll update the rest of the packages now - this was just a quick POC to get your feedback.
Do you think it warrants publishing new package versions immediately or can this wait a few days/weeks for the next patch of VS?

**Comment by BlythMeister** [comment]
Based on the speed of Vs updates, I rekon it could wait. (Also means the numbering of your packages stays "nice"

**Comment by jberezanski** [comment]
It's on master now. I think I'll keep this issue open until VS gets updated and the packages are pushed.

**Comment by Goru1890** [comment]
My installations fails when I use flags like `--add`, `--lang` or `--local`


`choco.exe install -y visualstudio2019buildtools --package-parameters "--add  Microsoft.VisualStudio.Component.VC.CoreBuildTools	 --includeRecommended --includeOptional --passive --lang En-us"
`

**Comment by jberezanski** [comment]
@Goru1890 your problem deserves a separate issue. This one deals with --includeRecommended / --includeOptional and I consider it resolved.

**Comment by ChrisMcKee** [comment]
@Goru1890 if, like me you're using it in packer, then you're better off using the "export config" option in the vs installer to get a .vsconfig , write it to disk and install from the online installer like this

```
$vsconfigfile = @"
{
  "version": "1.0",
  "components": [
    "Microsoft.Component.ClickOnce.MSBuild",
    "Microsoft.Component.MSBuild",
    "Microsoft.Net.Component.3.5.DeveloperTools",
    "Microsoft.Net.Component.4.5.1.TargetingPack",
    "Microsoft.Net.Component.4.5.2.TargetingPack",
    "Microsoft.Net.Component.4.5.TargetingPack",
    "Microsoft.Net.Component.4.6.1.TargetingPack",
    "Microsoft.Net.Component.4.6.2.TargetingPack",
    "Microsoft.Net.Component.4.6.TargetingPack",
    "Microsoft.Net.Component.4.7.1.TargetingPack",
    "Microsoft.Net.Component.4.7.2.TargetingPack",
    "Microsoft.Net.Component.4.7.TargetingPack",
    "Microsoft.Net.Component.4.8.SDK",
    "Microsoft.Net.Component.4.8.TargetingPack",
    "Microsoft.Net.Component.4.TargetingPack",
    "Microsoft.Net.ComponentGroup.4.6.2.DeveloperTools",
    "Microsoft.Net.ComponentGroup.4.7.1.DeveloperTools",
    "Microsoft.Net.ComponentGroup.4.8.DeveloperTools",
    "Microsoft.Net.ComponentGroup.DevelopmentPrerequisites",
    "Microsoft.Net.ComponentGroup.TargetingPacks.Common",
    "Microsoft.Net.Core.Component.SDK.2.1",
    "Microsoft.NetCore.Component.Runtime.3.1",
    "Microsoft.NetCore.Component.Runtime.5.0",
    "Microsoft.NetCore.Component.SDK",
    "Microsoft.VisualStudio.Component.AspNet45",
    "Microsoft.VisualStudio.Component.Azure.AuthoringTools",
    "Microsoft.VisualStudio.Component.Azure.ClientLibs",
    "Microsoft.VisualStudio.Component.Azure.Waverton.BuildTools",
    "Microsoft.VisualStudio.Component.CoreBuildTools",
    "Microsoft.VisualStudio.Component.DockerTools.BuildTools",
    "Microsoft.VisualStudio.Component.NuGet.BuildTools",
    "Microsoft.VisualStudio.Component.Roslyn.Compiler",
    "Microsoft.VisualStudio.Component.TestTools.BuildTools",
    "Microsoft.VisualStudio.Component.TypeScript.4.1",
    "Microsoft.VisualStudio.Component.WebDeploy",
    "Microsoft.VisualStudio.Wcf.BuildTools.ComponentGroup",
    "Microsoft.VisualStudio.Web.BuildTools.ComponentGroup",
    "Microsoft.VisualStudio.Workload.AzureBuildTools",
    "Microsoft.VisualStudio.Workload.MSBuildTools",
    "Microsoft.VisualStudio.Workload.WebBuildTools"
  ]
}
"@
$vsconfigfile|Out-File -FilePath C:\.vsconfig -Encoding ascii -Force

$vsPath = "$env:TEMP\vs_buildtools.exe"
(New-Object Net.WebClient).DownloadFile('https://aka.ms/vs/16/release/vs_buildtools.exe', $vsPath)

Start-Process -NoNewWindow -Wait $vsPath -ArgumentList '--wait --quiet --norestart --nocache --config c:\.vsconfig'
if (!$?) {
  Write-Host "Failed to install Visual Studio tools"
  Exit 1
}
```

At least you know what you're installing.
